File: mobile/src/crypto/ratchet.ts
/**
 * Shadow — Double Ratchet Algorithm (TypeScript)
 *
 * A complete implementation of the Signal Double Ratchet Algorithm:
 *   https://signal.org/docs/specifications/doubleratchet/
 *
 * Primitives:
 *   X25519       — Diffie-Hellman (@noble/curves/x25519)
 *   HKDF-SHA256  — KDF root chain (@noble/hashes/hkdf)
 *   HMAC-SHA256  — KDF chain keys (@noble/hashes/hmac)
 *   AES-256-GCM  — AEAD message encryption (Web Crypto SubtleCrypto)
 *
 * Public API:
 *   generateDH()
 *   dh(privHex, pubHex)
 *   kdfRk(rootKey, dhOut)            → [newRk, chainKey]
 *   kdfCk(chainKey)                  → [newCk, messageKey]
 *   aeadEncrypt(mk, pt, aad)         → Promise<Uint8Array>   (IV || ciphertext)
 *   aeadDecrypt(mk, ct, aad)         → Promise<Uint8Array>   (plaintext)
 *   serializeHeader / deserializeHeader
 *   ratchetInitAlice(sk, bobDhPub)   → RatchetState
 *   ratchetInitBob(sk, spkPriv, spkPub) → RatchetState
 *   ratchetEncrypt(state, plaintext, ad)
 *   ratchetDecrypt(state, header, ct, ad)
 */

import { x25519 }    from '@noble/curves/x25519';
import { sha256 }    from '@noble/hashes/sha256';
import { hkdf }      from '@noble/hashes/hkdf';
import { hmac }      from '@noble/hashes/hmac';
import { randomBytes } from '@noble/hashes/utils';
import { fromHex, toHex } from './identity';

// ─── Constants ────────────────────────────────────────────────────────────────

/** Maximum number of skipped message keys to keep. Prevents DoS. */
const MAX_SKIP = 1000;

const HKDF_INFO_RK  = new TextEncoder().encode('ShadowRootKey');
const HMAC_CK_CONST = new Uint8Array([0x01]);  // constant for chain key KDF
const HMAC_MK_CONST = new Uint8Array([0x02]);  // constant for message key KDF

/** AES-GCM nonce size (bytes) */
const GCM_NONCE_LEN = 12;

// ─── Types ────────────────────────────────────────────────────────────────────

/**
 * Complete state of a Double Ratchet session.
 * All keys are stored as lowercase hex for JSON serialisability.
 */
export interface RatchetState {
  /** Sending DH keypair — private key (hex) */
  dhsPriv:   string;
  /** Sending DH keypair — public key (hex) */
  dhsPub:    string;
  /** Receiving DH public key, or null before any message received */
  dhr:       string | null;
  /** Root key (32 bytes, hex) */
  rk:        string;
  /** Sending chain key (hex), or null if no sending chain yet */
  cks:       string | null;
  /** Receiving chain key (hex), or null if no receiving chain yet */
  ckr:       string | null;
  /** Send message number */
  ns:        number;
  /** Receive message number */
  nr:        number;
  /** Number of messages in previous sending chain */
  pn:        number;
  /**
   * Skipped message keys, keyed as "<dh_pub_hex>:<message_number>".
   * Values are message keys (hex).
   */
  mkSkipped: Record<string, string>;
}

export interface Header {
  /** Sender's current DH ratchet public key */
  dh: Uint8Array;
  /** Number of messages in previous sending chain */
  pn: number;
  /** Message number in current sending chain */
  n:  number;
}

export interface EncryptResult {
  /** Serialised 40-byte header */
  header:     Uint8Array;
  /** Promise resolving to (IV || GCM ciphertext) */
  ciphertext: Promise<Uint8Array>;
  /** Updated ratchet state (must be persisted before next operation) */
  newState:   RatchetState;
}

export interface DecryptResult {
  plaintext: string;
  newState:  RatchetState;
}

// ─── DH primitives ────────────────────────────────────────────────────────────

/**
 * Generate a fresh X25519 DH keypair.
 * Returns [privHex, pubHex].
 */
export function generateDH(): [string, string] {
  const priv = randomBytes(32);
  const pub  = x25519.getPublicKey(priv);
  return [toHex(priv), toHex(pub)];
}

/**
 * Perform X25519 Diffie-Hellman.
 * Returns the 32-byte shared secret.
 */
export function dh(privHex: string, pubHex: string): Uint8Array {
  return x25519.getSharedSecret(fromHex(privHex), fromHex(pubHex));
}

// ─── KDF functions ────────────────────────────────────────────────────────────

/**
 * Root key KDF. Derives a new root key and chain key from a DH output.
 *
 * RK', CK = HKDF-SHA256(RK, dh_out, 'ShadowRootKey', L=64)
 * Returns [newRootKey (32 bytes), chainKey (32 bytes)].
 */
export function kdfRk(
  rootKey: Uint8Array,
  dhOut:   Uint8Array,
): [Uint8Array, Uint8Array] {
  // HKDF: IKM=dhOut, salt=rootKey, info=HKDF_INFO_RK, length=64
  const out = hkdf(sha256, dhOut, rootKey, HKDF_INFO_RK, 64);
  return [out.slice(0, 32), out.slice(32, 64)];
}

/**
 * Chain key KDF. Derives a new chain key and message key.
 *
 * newCK = HMAC-SHA256(CK, 0x01)
 * MK    = HMAC-SHA256(CK, 0x02)
 *
 * Returns [newChainKey (32 bytes), messageKey (32 bytes)].
 */
export function kdfCk(chainKey: Uint8Array): [Uint8Array, Uint8Array] {
  const newCk = hmac(sha256, chainKey, HMAC_CK_CONST);
  const mk    = hmac(sha256, chainKey, HMAC_MK_CONST);
  return [newCk, mk];
}

// ─── AEAD: AES-256-GCM ───────────────────────────────────────────────────────

/**
 * Encrypt plaintext with AES-256-GCM.
 *
 * Output format: [12-byte IV] || [GCM ciphertext + 16-byte auth tag]
 *
 * @param messageKey — 32-byte AES key
 * @param plaintext  — plaintext bytes
 * @param aad        — additional authenticated data (not encrypted)
 */
export async function aeadEncrypt(
  messageKey: Uint8Array,
  plaintext:  Uint8Array,
  aad:        Uint8Array,
): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey(
    'raw',
    messageKey,
    { name: 'AES-GCM' },
    false,
    ['encrypt'],
  );
  const iv = randomBytes(GCM_NONCE_LEN);
  const ct = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv, additionalData: aad },
    key,
    plaintext,
  );
  // Prepend IV so receiver can extract it
  const result = new Uint8Array(GCM_NONCE_LEN + ct.byteLength);
  result.set(iv, 0);
  result.set(new Uint8Array(ct), GCM_NONCE_LEN);
  return result;
}

/**
 * Decrypt AES-256-GCM ciphertext.
 *
 * @param messageKey — 32-byte AES key
 * @param ciphertext — [12-byte IV] || [GCM ciphertext]
 * @param aad        — additional authenticated data (must match encrypt call)
 * @throws           DecryptionError if authentication fails
 */
export async function aeadDecrypt(
  messageKey: Uint8Array,
  ciphertext: Uint8Array,
  aad:        Uint8Array,
): Promise<Uint8Array> {
  if (ciphertext.length < GCM_NONCE_LEN + 16) {
    throw new Error('aeadDecrypt: ciphertext too short');
  }
  const key = await crypto.subtle.importKey(
    'raw',
    messageKey,
    { name: 'AES-GCM' },
    false,
    ['decrypt'],
  );
  const iv = ciphertext.slice(0, GCM_NONCE_LEN);
  const ct = ciphertext.slice(GCM_NONCE_LEN);
  const pt = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv, additionalData: aad },
    key,
    ct,
  );
  return new Uint8Array(pt);
}

// ─── Header serialisation ─────────────────────────────────────────────────────

/**
 * Serialise a ratchet header to a fixed 40-byte representation:
 *   bytes  0–31 : DH public key (32 bytes)
 *   bytes 32–35 : pn (uint32 big-endian)
 *   bytes 36–39 : n  (uint32 big-endian)
 */
export function serializeHeader(h: Header): Uint8Array {
  const out = new Uint8Array(40);
  out.set(h.dh, 0);
  const view = new DataView(out.buffer);
  view.setUint32(32, h.pn, false);  // big-endian
  view.setUint32(36, h.n,  false);
  return out;
}

/** Deserialise a 40-byte header. */
export function deserializeHeader(data: Uint8Array): Header {
  if (data.length < 40) {
    throw new RangeError(`deserializeHeader: expected 40 bytes, got ${data.length}`);
  }
  const dh   = data.slice(0, 32);
  const view = new DataView(data.buffer, data.byteOffset);
  const pn   = view.getUint32(32, false);
  const n    = view.getUint32(36, false);
  return { dh, pn, n };
}

/**
 * Build the AEAD additional data: AD || 4-byte header-length || header-bytes.
 * This binds ciphertexts to a specific header, preventing splice attacks.
 */
export function concatAD(ad: Uint8Array, header: Header): Uint8Array {
  const hdrBytes = serializeHeader(header);
  const lenBuf   = new Uint8Array(4);
  new DataView(lenBuf.buffer).setUint32(0, hdrBytes.length, false);
  const out = new Uint8Array(ad.length + 4 + hdrBytes.length);
  out.set(ad,       0);
  out.set(lenBuf,   ad.length);
  out.set(hdrBytes, ad.length + 4);
  return out;
}

// ─── Ratchet initialisation ───────────────────────────────────────────────────

/**
 * Initialise Alice's (sender's) ratchet state after a completed X3DH exchange.
 *
 * @param sk       — 32-byte shared secret from X3DH
 * @param bobDhPub — Bob's signed prekey public (hex); Alice uses it as dhr
 */
export function ratchetInitAlice(
  sk:       Uint8Array,
  bobDhPub: string,
): RatchetState {
  const [dhsPriv, dhsPub] = generateDH();
  const dhOut             = dh(dhsPriv, bobDhPub);
  const [rk, cks]         = kdfRk(sk, dhOut);

  return {
    dhsPriv, dhsPub,
    dhr: bobDhPub,
    rk:  toHex(rk),
    cks: toHex(cks),
    ckr: null,
    ns: 0, nr: 0, pn: 0,
    mkSkipped: {},
  };
}

/**
 * Initialise Bob's (receiver's) ratchet state after receiving an X3DH message.
 *
 * @param sk      — 32-byte shared secret from X3DH
 * @param spkPriv — Bob's signed prekey private key (hex)
 * @param spkPub  — Bob's signed prekey public key (hex)
 */
export function ratchetInitBob(
  sk:      Uint8Array,
  spkPriv: string,
  spkPub:  string,
): RatchetState {
  return {
    dhsPriv: spkPriv,
    dhsPub:  spkPub,
    dhr:     null,
    rk:      toHex(sk),
    cks:     null,
    ckr:     null,
    ns: 0, nr: 0, pn: 0,
    mkSkipped: {},
  };
}

// ─── Encrypt ──────────────────────────────────────────────────────────────────

/**
 * Encrypt a plaintext message using the current sending chain.
 *
 * The header and updated state are returned synchronously (important: the
 * caller MUST persist `newState` before awaiting `ciphertext`).
 * The ciphertext is returned as a Promise because AES-GCM is async.
 *
 * @throws if there is no active sending chain (cks is null)
 */
export function ratchetEncrypt(
  state:     RatchetState,
  plaintext: string,
  ad:        Uint8Array,
): EncryptResult {
  if (!state.cks) {
    throw new Error('ratchetEncrypt: no sending chain key (session not initialised)');
  }

  // Deep-clone state so caller's original is not mutated
  const s: RatchetState = {
    ...state,
    mkSkipped: { ...state.mkSkipped },
  };

  // Advance the sending chain
  const [newCk, mk] = kdfCk(fromHex(s.cks!));
  s.cks = toHex(newCk);

  const header: Header = { dh: fromHex(s.dhsPub), pn: s.pn, n: s.ns };
  s.ns += 1;

  const aad        = concatAD(ad, header);
  const ptBytes    = new TextEncoder().encode(plaintext);
  const ciphertext = aeadEncrypt(mk, ptBytes, aad);

  return { header: serializeHeader(header), ciphertext, newState: s };
}

// ─── Decrypt ──────────────────────────────────────────────────────────────────

/**
 * Decrypt a received message, advancing the ratchet as needed.
 *
 * Handles:
 *   - Skipped messages (out-of-order delivery)
 *   - DH ratchet steps (new sender ratchet key)
 *
 * @throws if decryption fails (wrong key, tampered ciphertext, etc.)
 */
export async function ratchetDecrypt(
  state:       RatchetState,
  headerBytes: Uint8Array,
  ciphertext:  Uint8Array,
  ad:          Uint8Array,
): Promise<DecryptResult> {
  // Deep-clone state
  const s: RatchetState = {
    ...state,
    mkSkipped: { ...state.mkSkipped },
  };

  const header = deserializeHeader(headerBytes);
  const dhHex  = toHex(header.dh);

  // ── Try previously skipped message keys ────────────────────────────────
  const skippedKey = `${dhHex}:${header.n}`;
  if (s.mkSkipped[skippedKey]) {
    const mk = fromHex(s.mkSkipped[skippedKey]);
    delete s.mkSkipped[skippedKey];
    const aad = concatAD(ad, header);
    const pt  = await aeadDecrypt(mk, ciphertext, aad);
    return { plaintext: new TextDecoder().decode(pt), newState: s };
  }

  // ── DH ratchet step if needed ──────────────────────────────────────────
  if (s.dhr === null || dhHex !== s.dhr) {
    skipMessageKeys(s, header.pn);
    dhRatchetStep(s, header);
  }

  // ── Advance receive chain ──────────────────────────────────────────────
  skipMessageKeys(s, header.n);

  if (!s.ckr) {
    throw new Error('ratchetDecrypt: no receiving chain key after DH ratchet');
  }

  const [newCk, mk] = kdfCk(fromHex(s.ckr));
  s.ckr  = toHex(newCk);
  s.nr  += 1;

  const aad = concatAD(ad, header);
  const pt  = await aeadDecrypt(mk, ciphertext, aad);
  return { plaintext: new TextDecoder().decode(pt), newState: s };
}

// ─── Private helpers ──────────────────────────────────────────────────────────

/**
 * Store message keys for all skipped messages up to `until`.
 * Mutates `s` in place.
 */
function skipMessageKeys(s: RatchetState, until: number): void {
  if (s.nr + MAX_SKIP < until) {
    throw new Error(
      `ratchetDecrypt: too many skipped messages (${until - s.nr} > ${MAX_SKIP})`,
    );
  }
  while (s.ckr && s.nr < until) {
    const [newCk, mk] = kdfCk(fromHex(s.ckr));
    s.ckr = toHex(newCk);
    s.mkSkipped[`${s.dhr}:${s.nr}`] = toHex(mk);
    s.nr += 1;
  }
}

/**
 * Perform a DH ratchet step on receiving a new sender DH key.
 * Derives a new receiving chain key, then a new sending chain key.
 * Mutates `s` in place.
 */
function dhRatchetStep(s: RatchetState, header: Header): void {
  s.pn  = s.ns;
  s.ns  = 0;
  s.nr  = 0;
  s.dhr = toHex(header.dh);

  // Receiving ratchet: DH(ours, their new key) → new RK + CKr
  const out1      = dh(s.dhsPriv, s.dhr);
  const [rk1, ckr] = kdfRk(fromHex(s.rk), out1);
  s.rk  = toHex(rk1);
  s.ckr = toHex(ckr);

  // Generate a new sending DH keypair
  const [newPriv, newPub] = generateDH();
  s.dhsPriv = newPriv;
  s.dhsPub  = newPub;

  // Sending ratchet: DH(new ours, their new key) → new RK + CKs
  const out2      = dh(s.dhsPriv, s.dhr);
  const [rk2, cks] = kdfRk(fromHex(s.rk), out2);
  s.rk  = toHex(rk2);
  s.cks = toHex(cks);
}

File: core/ratchet.py
"""
Shadow — Double Ratchet Protocol
Reference: https://signal.org/docs/specifications/doubleratchet/

Primitives:
  - X25519       — Diffie-Hellman
  - HKDF-SHA256  — key derivation
  - AES-256-GCM  — AEAD encryption
  - HMAC-SHA256  — symmetric chain ratchet
"""

import os
import hmac
import hashlib
import struct
from dataclasses import dataclass, field
from typing import Optional

from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.serialization import (
    Encoding, PublicFormat, PrivateFormat, NoEncryption
)
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


# --------------------------------------------------------------------------- #
# Constants (Signal spec §2.2)
# --------------------------------------------------------------------------- #

MAX_SKIP = 1000          # maximum skipped message keys to store
HKDF_INFO_RK  = b"ShadowRootKey"
HKDF_INFO_MSG = b"ShadowMessageKey"
HMAC_CK_CONST = b"\x01"   # derive next chain key
HMAC_MK_CONST = b"\x02"   # derive message key


# --------------------------------------------------------------------------- #
# DH helpers
# All key material is stored as raw bytes (32 bytes) for serializability.
# X25519 key objects are created transiently for DH operations only.
# --------------------------------------------------------------------------- #

def generate_dh() -> tuple[bytes, bytes]:
    """Return (private_bytes, public_bytes) X25519 pair — both 32 raw bytes."""
    priv = X25519PrivateKey.generate()
    priv_raw = priv.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption())
    pub_raw  = priv.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
    return priv_raw, pub_raw


def dh(private_bytes: bytes, public_bytes: bytes) -> bytes:
    """X25519 DH: private_bytes × public_bytes → 32-byte shared secret."""
    priv = X25519PrivateKey.from_private_bytes(private_bytes)
    peer = X25519PublicKey.from_public_bytes(public_bytes)
    return priv.exchange(peer)


# --------------------------------------------------------------------------- #
# KDF functions
# --------------------------------------------------------------------------- #

def kdf_rk(root_key: bytes, dh_out: bytes) -> tuple[bytes, bytes]:
    """
    KDF_RK(rk, dh_out) → (new_root_key, chain_key)
    Uses HKDF-SHA256 with the root key as salt.
    """
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=64,
        salt=root_key,
        info=HKDF_INFO_RK,
    )
    output = hkdf.derive(dh_out)
    return output[:32], output[32:]


def kdf_ck(chain_key: bytes) -> tuple[bytes, bytes]:
    """
    KDF_CK(ck) → (new_chain_key, message_key)
    Uses HMAC-SHA256 with Signal spec constants 0x01 / 0x02.
    """
    new_ck = hmac.new(chain_key, HMAC_CK_CONST, hashlib.sha256).digest()
    mk     = hmac.new(chain_key, HMAC_MK_CONST, hashlib.sha256).digest()
    return new_ck, mk


# --------------------------------------------------------------------------- #
# AEAD: AES-256-GCM
# --------------------------------------------------------------------------- #

def encrypt(message_key: bytes, plaintext: bytes, associated_data: bytes) -> bytes:
    """AES-256-GCM encrypt. Returns nonce || ciphertext+tag."""
    nonce = os.urandom(12)
    aead  = AESGCM(message_key)
    ct    = aead.encrypt(nonce, plaintext, associated_data)
    return nonce + ct


def decrypt(message_key: bytes, ciphertext: bytes, associated_data: bytes) -> bytes:
    """AES-256-GCM decrypt. Expects nonce || ciphertext+tag."""
    nonce = ciphertext[:12]
    ct    = ciphertext[12:]
    aead  = AESGCM(message_key)
    return aead.decrypt(nonce, ct, associated_data)


# --------------------------------------------------------------------------- #
# Message header
# --------------------------------------------------------------------------- #

@dataclass
class Header:
    dh: bytes       # sender's current ratchet public key (32 bytes)
    pn: int         # number of messages in previous sending chain
    n:  int         # message number in current sending chain

    def serialize(self) -> bytes:
        """32-byte DH key || 4-byte PN || 4-byte N (big-endian)."""
        return self.dh + struct.pack(">II", self.pn, self.n)

    @classmethod
    def deserialize(cls, data: bytes) -> "Header":
        if len(data) < 40:
            raise ValueError(f"Header too short: {len(data)} bytes (need 40)")
        dh = data[:32]
        pn, n = struct.unpack(">II", data[32:40])
        return cls(dh=dh, pn=pn, n=n)

    def __len__(self):
        return 40


def concat_ad(associated_data: bytes, header: Header) -> bytes:
    """Bind the header to the associated data so it is authenticated."""
    hdr_bytes = header.serialize()
    return associated_data + struct.pack(">I", len(hdr_bytes)) + hdr_bytes


# --------------------------------------------------------------------------- #
# Double Ratchet state
# All byte fields are plain bytes so the state is fully copyable / serializable.
# --------------------------------------------------------------------------- #

@dataclass
class RatchetState:
    DHs:       tuple                    # (private_bytes, public_bytes) — 32 raw bytes each
    DHr:       Optional[bytes]          # their current ratchet public key (32 raw bytes)
    RK:        bytes                    # 32-byte root key
    CKs:       Optional[bytes]          # sending chain key
    CKr:       Optional[bytes]          # receiving chain key
    Ns:        int = 0                  # sending message counter
    Nr:        int = 0                  # receiving message counter
    PN:        int = 0                  # previous chain message count
    MKSKIPPED: dict = field(default_factory=dict)  # {(dh_pub, n): message_key}


# --------------------------------------------------------------------------- #
# Ratchet initialisation (Signal spec §3.3)
# --------------------------------------------------------------------------- #

def ratchet_init_alice(SK: bytes, bob_dh_public: bytes) -> RatchetState:
    """
    Alice initialises after X3DH.
    SK             — shared secret from X3DH
    bob_dh_public  — Bob's signed prekey public bytes (initial ratchet key)
    """
    dhs_priv, dhs_pub = generate_dh()
    rk, cks = kdf_rk(SK, dh(dhs_priv, bob_dh_public))
    return RatchetState(
        DHs=(dhs_priv, dhs_pub),
        DHr=bob_dh_public,
        RK=rk,
        CKs=cks,
        CKr=None,
    )


def ratchet_init_bob(SK: bytes, bob_dh_pair: tuple) -> RatchetState:
    """
    Bob initialises after X3DH.
    SK           — shared secret from X3DH
    bob_dh_pair  — (private_bytes, public_bytes) Bob's signed prekey pair
    """
    return RatchetState(
        DHs=bob_dh_pair,
        DHr=None,
        RK=SK,
        CKs=None,
        CKr=None,
    )


# --------------------------------------------------------------------------- #
# Encrypt (Signal spec §3.4)
# --------------------------------------------------------------------------- #

def ratchet_encrypt(state: RatchetState, plaintext: bytes, AD: bytes) -> tuple[Header, bytes]:
    """
    Encrypt plaintext using state.
    Returns (header, ciphertext).
    Advances the sending chain.
    """
    state.CKs, mk = kdf_ck(state.CKs)
    header = Header(dh=state.DHs[1], pn=state.PN, n=state.Ns)
    state.Ns += 1
    ct = encrypt(mk, plaintext, concat_ad(AD, header))
    return header, ct


# --------------------------------------------------------------------------- #
# Decrypt helpers (Signal spec §3.5)
# --------------------------------------------------------------------------- #

def _try_skipped_message_keys(
    state: RatchetState,
    header: Header,
    ciphertext: bytes,
    AD: bytes,
) -> Optional[bytes]:
    key = (bytes(header.dh), header.n)
    if key in state.MKSKIPPED:
        mk = state.MKSKIPPED.pop(key)
        return decrypt(mk, ciphertext, concat_ad(AD, header))
    return None


def _skip_message_keys(state: RatchetState, until: int) -> None:
    if state.Nr + MAX_SKIP < until:
        raise ValueError(f"Too many skipped messages: {until - state.Nr}")
    if state.CKr is not None:
        while state.Nr < until:
            state.CKr, mk = kdf_ck(state.CKr)
            state.MKSKIPPED[(bytes(state.DHr), state.Nr)] = mk
            state.Nr += 1


def _dh_ratchet(state: RatchetState, header: Header) -> None:
    """Perform a DH ratchet step on receipt of a new sender ratchet key."""
    state.PN = state.Ns
    state.Ns = 0
    state.Nr = 0
    state.DHr = header.dh
    state.RK, state.CKr = kdf_rk(state.RK, dh(state.DHs[0], state.DHr))
    state.DHs = generate_dh()
    state.RK, state.CKs = kdf_rk(state.RK, dh(state.DHs[0], state.DHr))


def ratchet_decrypt(state: RatchetState, header: Header, ciphertext: bytes, AD: bytes) -> bytes:
    """
    Decrypt a received message.
    Handles out-of-order delivery via skipped message key store.
    """
    # 1. Check skipped keys first
    plaintext = _try_skipped_message_keys(state, header, ciphertext, AD)
    if plaintext is not None:
        return plaintext

    # 2. DH ratchet step if sender has rotated their key
    if state.DHr is None or bytes(header.dh) != bytes(state.DHr):
        _skip_message_keys(state, header.pn)
        _dh_ratchet(state, header)

    # 3. Skip within current chain (out-of-order)
    _skip_message_keys(state, header.n)

    # 4. Decrypt
    state.CKr, mk = kdf_ck(state.CKr)
    state.Nr += 1
    return decrypt(mk, ciphertext, concat_ad(AD, header))
